### PR TITLE
Added support for parsing multi-docket NEFs

### PR DESCRIPTION
**Pull request by albertisfu**
This PR adds support for parsing NEFs that mention multiple dockets for the same PACER document/attachments like the example in https://github.com/freelawproject/courtlistener/issues/2270

Added new examples for multi-docket NEFS: nysd_4.json, nysd_3.json, and jpml_1.json
And updated all the current examples to match the new data response structure for the `NotificationEmail` report.

Now we have a new key "dockets" that contains all the dockets mentioned in the notification.

Since we only have `text/html` multi-docket NEFs  examples, right now we don't support parsing text/plain NEFs  or multi-docket NDAS multi-docket (in case they exist), so I added an exception to fail it loud when we'll get a notification for `text/plain` multi-docket NEFs of multi-docket NDAs so we can get the example and add support for them.

Once this is merged I'll add a version bump and update the CL PR.

Let me know what you think.

### CODE REVIEW

**Comment by johnhawkinson** [comment]
> Since we only have `text/html` multi-docket NEFs examples, right now we don't support parsing text/plain NEFs or multi-docket NDAS multi-docket (in case they exist),

They do and they do not. When a docket entry is "spread" (the ECF technical term across both DC and Appellate) to multiple dockets, CMECF Appellate generates  an NDAs, which includes at the end of the docket text a bracketed list of all cases to which the entry is spread. Regrettably my email setup currently does Message-ID -based duplicate suppression, so I am not 100% what actually happens, but I think that two near-identical NDAs are generated, possibly with the same Message-ID. I don't seem to have multiple copies of the NDAs for each case, although sometimes when an NDA is regenerated, I seem to have copies for different cases.

Interestingly the list of docket numbers is ordered differently in each case, listing the case for which the notification is generated first. Another related difference is CMECF Appellate allows non-Court filers to spread a docket entry among multiple cases for which a relation has already been defined by the Court. CMECF DC does not permit that for non-Court users.

Here's an example pair of this confusion:


<hr>

```smtp
Date: Wed, 6 Jul 2022 10:55:37 -0400 (EDT)                                                               
From: CA01NoticeDocketActivity@ca1.uscourts.gov                                                          
To: jhawk@alum.mit.edu                                                                                   
Subject: 20-1119 Pereira Brito, et al v. Garland, et al "Court Order"                                    
X-REMOTE-IP: 156.119.190.135                                                                             
Message-ID: <972927161.312.1657119337647@ca1db.ca1.gtwy.dcn>                                             
```

<p><strong>
***NOTE TO PUBLIC ACCESS USERS*** Judicial Conference of the United 
States policy permits attorneys of record and parties in a case 
(including pro se litigants) to receive one free electronic copy of all 
documents filed electronically, if receipt is required by law or 
directed by the filer.  PACER access fees apply to all other users.  To 
avoid later charges, download a copy of each document during this first 
viewing.</strong><br>
</p><p></p><center><strong>United States Court of Appeals for the First Circuit</strong></center>
<p><strong>Notice of Docket Activity</strong><br>
<br>
The following transaction was entered on 07/06/2022 at 10:55:12 AM Eastern Daylight Time and filed on 07/06/2022
</p>

Case Name: | Pereira Brito, et al v. Garland, et al
-- | --
Case Number: | 20-1119
Document(s): | Document(s)

<br>
<p><strong>Docket Text:</strong><br>
ORDER entered by David J. Barron, Chief Appellate Judge; Sandra L. 
Lynch, Appellate Judge; Kermit V. Lipez, Appellate Judge; Rogeriee 
Thompson, Appellate Judge; William J. Kayatta, Jr., Appellate Judge and 
Gustavo A. Gelpi, Jr., Appellate Judge: The petition for rehearing 
having been denied by the panel of judges who decided the case, and the 
petition for rehearing en banc having been submitted to the active 
judges of this court and a majority of the judges not having voted that 
the case be heard en banc, it is ordered that the petition for rehearing
 and the petition for rehearing en banc filed by 
petitioners-appellants/cross-<wbr>appellees Gilberto Pereira <span class="il">Brito</span>, Florentin Avila Lucas, and Jacky Celicourt be denied. [20-1119, 20-1037] (JMK)<br>
<br>
<strong>Notice will be electronically mailed to:</strong><br><br>
Gilles R. Bissonnette<br>
…

<hr>

```smtp
Date: Wed, 6 Jul 2022 11:00:19 -0400 (EDT)                                                               
From: CA01NoticeDocketActivity@ca1.uscourts.gov                                                          
To: jhawk@alum.mit.edu                                                                                   
Subject: 20-1037 Pereira Brito, et al v. Garland, et al "Court Order"                                    
X-REMOTE-IP: 156.119.190.135                                                                             
Message-ID: <922819282.363.1657119619515@ca1db.ca1.gtwy.dcn>                                             

```

<p><strong>
***NOTE TO PUBLIC ACCESS USERS*** Judicial Conference of the United 
States policy permits attorneys of record and parties in a case 
(including pro se litigants) to receive one free electronic copy of all 
documents filed electronically, if receipt is required by law or 
directed by the filer.  PACER access fees apply to all other users.  To 
avoid later charges, download a copy of each document during this first 
viewing.</strong><br>
</p><p></p><center><strong>United States Court of Appeals for the First Circuit</strong></center>
<p><strong>Notice of Docket Activity</strong><br>
<br>
The following transaction was entered on 07/06/2022 at 11:00:00 AM Eastern Daylight Time and filed on 07/06/2022
</p>

Case Name: | Pereira Brito, et al v. Garland, et al
-- | --
Case Number: | 20-1037
Document(s): | Document(s)

<br>
<p><strong>Docket Text:</strong><br>
ORDER entered by David J. Barron, Chief Appellate Judge; Sandra L. 
Lynch*, Appellate Judge; Kermit V. Lipez, Appellate Judge; Rogeriee 
Thompson, Appellate Judge; William J. Kayatta, Jr., Appellate Judge and 
Gustavo A. Gelpi, Jr., Appellate Judge: Pursuant to First Circuit 
Internal Operating Procedure X(C), the petition for rehearing en banc 
filed by respondents-appellees/cross-<wbr>appellants Merrick B. Garland,
 Timothy S. Robbins, Tae D. Johnson, Alejandro Mayorkas, Jean King, 
Antone Moniz, Yolanda Smith, Steven Souza, Christopher Brackett, and 
Lori Streeter has also been treated as a petition for rehearing before 
the original panel.  The petition for rehearing having been denied by 
the panel of judges who decided the case, and the petition for rehearing
 en banc having been submitted to the active judges of this court and a 
majority of the judges not having voted that the case be heard en banc, 
it is ordered that the petition for rehearing and petition for rehearing
 en banc be denied. *<b>LYNCH, <u>Circuit Judge</u></b>, dissents from the denial of en banc rehearing. [20-1037, 20-1119] (JMK)<br>
<br>
<strong>Notice will be electronically mailed to:</strong><br><br>
Gilles R. Bissonnette</p>***NOTE TO PUBLIC ACCESS USERS*** Judicial Conference of the United States policy permits attorneys of record and parties in a case (including pro se litigants) to receive one free electronic copy of all documents filed electronically, if receipt is required by law or directed by the filer. PACER access fees apply to all other users. To avoid later charges, download a copy of each document during this first viewing.

United States Court of Appeals for the First Circuit

Notice of Docket Activity

The following transaction was entered on 07/06/2022 at 11:00:00 AM Eastern Daylight Time and filed on 07/06/2022
Case Name: 	Pereira Brito, et al v. Garland, et al
Case Number:   	[20-1037](https://ecf.ca1.uscourts.gov/n/beam/servlet/TransportRoom?servlet=DocketReportFilter.jsp&caseId=46408)
Document(s): 	[Document(s)](https://ecf.ca1.uscourts.gov/docs1/00107895169?uid=adf1c171b021ae99)

Docket Text:
ORDER entered by David J. Barron, Chief Appellate Judge; Sandra L. Lynch*, Appellate Judge; Kermit V. Lipez, Appellate Judge; Rogeriee Thompson, Appellate Judge; William J. Kayatta, Jr., Appellate Judge and Gustavo A. Gelpi, Jr., Appellate Judge: Pursuant to First Circuit Internal Operating Procedure X(C), the petition for rehearing en banc filed by respondents-appellees/cross-appellants Merrick B. Garland, Timothy S. Robbins, Tae D. Johnson, Alejandro Mayorkas, Jean King, Antone Moniz, Yolanda Smith, Steven Souza, Christopher Brackett, and Lori Streeter has also been treated as a petition for rehearing before the original panel. The petition for rehearing having been denied by the panel of judges who decided the case, and the petition for rehearing en banc having been submitted to the active judges of this court and a majority of the judges not having voted that the case be heard en banc, it is ordered that the petition for rehearing and petition for rehearing en banc be denied. *LYNCH, Circuit Judge, dissents from the denial of en banc rehearing. [20-1037, 20-1119] (JMK)

Notice will be electronically mailed to:

Gilles R. Bissonnette<br>
…
<hr>




p.s.: I know, I know, you probably wanted the raw html. I decided to make the github version look pretty instead. Was there a way to do both?

**Comment by mlissner** [comment]
> I know, I know, you probably wanted the raw html. I decided to make the github version look pretty instead. Was there a way to do both?

I think you'd have to email/slack/etc us the HTML. IIRC, Github doesn't let you upload raw HTML for some reason (security, probably?).

Handling the rest looks pretty painful. I don't relish parsing out docket numbers from the Docket Text section, though I guess we should/could get an issue going to think about that.

File: juriscraper/pacer/email.py [hunk 1]
```diff
@@ -77,53 +91,63 @@ def _sibling_path(self, label):
...
+        Raise an exception if a potential multi-docket text/plain notification
+        is found so we can get the example and add support for it.
+
         :returns: Case name, cleaned and harmonized
         """
         email_body = self.tree.text_content()
         regex = r"Case Name:(.*)"
         find_case = re.findall(regex, email_body)
+        if len(find_case) > 1:
+            raise Exception(
```
Comment (mlissner) [inline juriscraper/pacer/email.py:122]:
```suggestion
            raise NotImplementedError(
```

File: juriscraper/pacer/email.py [hunk 2]
```diff
@@ -77,53 +91,63 @@ def _sibling_path(self, label):
...
+                f"Received a potential multi-docket text/plain notification, "
+                f"court: {self.court_id}"
```
Comment (mlissner) [inline juriscraper/pacer/email.py:124]:
```suggestion
                f"Received a potential multi-docket text/plain notification. "
                f"This is probably our chance to add support for it. "
                f"court: {self.court_id}"
```

File: juriscraper/pacer/email.py [hunk 3]
```diff
@@ -303,11 +331,54 @@ def _contains_attachments_plain(self) -> bool:
...
+            dockets.append(docket)
+        else:
+            dockets_table = self.tree.xpath(
+                "//table[contains(., 'Case Name:')]"
+            )
+            if self.appellate and len(dockets_table) > 1:
+                raise Exception(
+                    f"Received a potential multi-docket NDA notification, "
+                    f"court: {self.court_id}"
+                )
```
Comment (mlissner) [inline juriscraper/pacer/email.py:361]:
```suggestion
                raise NotImplementedError(
                    f"Received a potential multi-docket NDA notification. "
                    f"This is probably our chance to add support for it. "
                    f"court: {self.court_id}"
                )
```

**Review by mlissner — CHANGES_REQUESTED** [review]
This looks about right. A couple small comments, but that's it.

**Comment by albertisfu** [comment]
Thanks @mlissner I'll work on these changes.

@johnhawkinson thanks for the examples. 

> I think that two near-identical NDAs are generated, possibly with the same Message-ID. I don't seem to have multiple copies of the NDAs for each case

If there are duplicated NDAs and they are exactly the same (same docket number), I agree with Mike, that would be a different issue to think about how to handle them.  If they are not exactly the same and each of them mentions a different docket number I think we can handle them right now.

**Comment by johnhawkinson** [comment]
> If there are duplicated NDAs and they are exactly the same (same docket number), I agree with Mike, that would be a different issue to think about how to handle them. If they are not exactly the same and each of them mentions a different docket number I think we can handle them right now.

If "same docket number" means "same docket number in the Subject: line and in the Case Number:" field, then obviously no they are not exactly the same, as strongly suggested by my examples with different Subject lines. If "same docket number" means "same docket text save for the bracketed enumeration of docket numbers at the end," then I think they are.

**Comment by albertisfu** [comment]
@johnhawkinson yeah I meant in the `Case Number` field that is the one we use to extract the docket number. So yes if they are not the same in the notifications we're going to be able to handle them independently, adding the docket entry to each case accordingly.

File: juriscraper/pacer/email.py (hunk 1 above)
Comment (albertisfu) [reply juriscraper/pacer/email.py]: Changed!

File: juriscraper/pacer/email.py (hunk 2 above)
Comment (albertisfu) [reply juriscraper/pacer/email.py]: Text updated!

File: juriscraper/pacer/email.py (hunk 3 above)
Comment (albertisfu) [reply juriscraper/pacer/email.py]: Text updated!

**Review by mlissner — APPROVED** [review]

**Pull request merged**